docs/proposals/scalable-query-frontend.md

```diff
@@ -0,0 +1,267 @@
+
+Add a short backoff to the main frontend [processing loop](https://github.com/cortexproject/cortex/blob/8fb86155a7c7c155b8c4d31b91b267f9631b60ba/pkg/querier/frontend/frontend.go#L288-L331).  This would cause the frontend to briefly back off of any querier that was overloaded but continue to send queries to those that were capable of doing work.
+
+```go
+if current > desired {
+  zzz := (current - desired) * backoffDuration
+  zzz *= 1 + rand.Float64() * .1               // jitter
+  time.Sleep(zzz)
+}
+```
+
+Passing flow control information from the querier to the frontend would also open up additional future work for more sophisticated load balancing across queriers.  For example by simply comparing and choosing [the least congested of two](https://www.nginx.com/blog/nginx-power-of-two-choices-load-balancing-algorithm/) queriers we could dramatically improve how well work is distributed.
+
+### Increased Time To Failure
+
+#### Challenge
+
+Scaling the query frontend also increases the per tenant queue length by creating more queues.  This could result in increased latencies where failing fast (429) would have been preferred.
+
+The operator could reduce the queue length per query frontend in response to scaling out, but
+
+then they would run the risk of unnecessarily failing a request due to unbalanced distribution across query frontends.  Also, shorter queues run the risk of failing to properly service heavily sharded queries.
+
+Another concern is that a system with more queues will take longer to recover from an production event as it will have queued up more work.
+
+#### Proposal
+
+Currently we are not proposing any changes to alleviate this concern.  We believe this is solvable operationally.  This can be revisited as more information is gathered.
+
+### Querier Discovery Lag
+
+#### Challenge
+
+Queriers have a configurable parameter that controls how often they refresh their query frontend list.  The default value is 10 seconds.  After a new query frontend is added the average querier will take 5 seconds (after DNS is updated) to become aware of it and begin requesting queries from it.
+
+#### Proposal
+
+It is recommended to add a readiness/health check to the query frontend to prevent it from receiving queries while it is waiting for queriers to connect.   HTTP health checks are supported by [envoy](https://www.envoyproxy.io/learn/health-check), [k8s](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/), [nginx](https://docs.nginx.com/nginx/admin-guide/load-balancer/http-health-check/), and basically any commodity load balancer.  The query frontend would not indicate healthy on its health check until at least one querier had connected.
+
+
+### Dilutes Tenant Fairness
+
+#### Challenge
+
+Given `f` query frontends, `n` tenants and an average of `q` queries in the frontend per tenant.  The number of tenants per instance would be:
```

Comment:
I believe you're assuming that something is distributing requests equally across all query frontends; this assumption should be stated.
An alternative might be to distribute using consistent hashing on the tenant ID, which I think would change the outcome.

Reply:
> I believe you're assuming that something is distributing requests equally across all query frontends; this assumption should be stated.

I had noted it a bit below.  Moved it to the top for clarity to show it applied to all equations.

> An alternative might be to distribute using consistent hashing on the tenant ID, which I think would change the outcome.

This is interesting, but would require a tenant aware component in front of the query-frontends.  I wonder if there is a commodity load balancer (haproxy, nginx, envoy, whatever) that will do that kind of load balancing using an HTTP header.

Reply:
https://www.haproxy.com/blog/haproxys-load-balancing-algorithm-for-static-content-delivery-with-varnish/#hash-on-a-http-header

Reply:
I'm not sure how I feel about an approach like this.  It would definitely help with tenancy fairness, but I'd be concerned about one instance getting overloaded due to a very active tenant.  Also it would suffer from "noisy neighbor" issues where low volume tenants could land on the same frontend as high demand tenants and get reduced QoS due to bad luck.

Shuffle sharding helps deal with some of this, but would almost certainly require a custom implementation.